osmomath/decimal.go

```diff
@@ -552,23 +560,45 @@ func (d BigDec) MustFloat64() float64 {
 // Dec returns the osmomath.Dec representation of a BigDec.
 // Values in any additional decimal places are truncated.
 func (d BigDec) Dec() Dec {
-	precisionDiff := PrecisionBigDec - PrecisionDec
-	precisionFactor := new(big.Int).Exp(big.NewInt(10), big.NewInt(int64(precisionDiff)), nil)
+	return d.DecWithPrecision(PrecisionDec)
+}
 
-	if precisionDiff < 0 {
-		panic("invalid decimal precision")
+func (d BigDec) DecWithPrecision(precision int64) Dec {
+	var precisionFactor *big.Int
+	if precision >= PrecisionDec { // if provided precision is more than Dec's precision, assume conversion to native Dec (no precision lost)
```

Comment:
I think we should be failing if precision is > PrecisionDec.

Also, please provide a detailed method spec outlining all combos of inputs and outputs

Reply:
hey, made it panic if provided precision is wrong. Could you please elaborate on what you mean by "all combos of inputs and outputs"? I just added a small example to godoc of this function, if that is what you are referring to: https://github.com/osmosis-labs/osmosis/pull/6261/commits/7da117bfe09812f85a87cd78fb874ffab82afd2f

cc: @p0mvn

Reply:
If precision is greater than PrecisionDec => panic
Otherwise, perform the conversion to 18 decimals that will have a given precision

Reply:
Also, why is `precision` an `int64`? It cannot be negative. Please make it `uint64`

Reply:
Changed